### PR TITLE
osu! commands

### PR DESCRIPTION
**Pull request by jaxxibae**
# osu! user lookup (taiko, ctb, mania and osu!)

imagem
imagem
imagem

# osu! beatmap lookup (same modes)

imagem
imagem
imagem

# have the ability to register a osu! player to a Discord acct

it's not thru discord anymore, it's now a freaking connection! 

- Closes #54

### CODE REVIEW

**Comment by dpaiv0** [comment]
### RFR

File: src/locales/en-US/commands.json [hunk 1]
```diff
@@ -1017,6 +1017,50 @@
...
+  "osu": {
+    "commandDescription": "Looks up a player or a beatmap on osu!",
+    "commandUsage": "<type> <query>",
+    "subcommands": {
+      "player": {
+        "commandDescription": "Looks up a player on osu!",
+        "commandUsage": "[player name or id]",
+        "noPlayer": "You need to give me a valid osu! player name or ID.",
+        "level": "Level **{{number}}**",
+        "joinedAt": "Joined **{{date}}** ({{timeAgo}})",
```
Comment (LogGame-HU-DEaN) [inline src/locales/en-US/commands.json]: How about „Joined **on** **{{date}}** ({{timeAgo}})”?

Comment (dpaiv0) [reply src/locales/en-US/commands.json]: sure

**Comment by Lireoy** [comment]
@dpaiv0 what's the theory between the register function? an osu account info to be linked to a discord account? or what? can you pls clarify?

**Comment by dpaiv0** [comment]
@Lireoy imagine if you're me: you're consistently watching your data on osu!, so that you can keep track of your growth, but you don't want to type `s!osu u [ PillGP ]` every time, so you register / save your name with `s!osu r [ PillGP ]`, it saves your osu! ID (not username, because it's changeable), and the next time you type `s!osu u`, it will return (if existant) your saved user's data

File: src/commands/games/osu/register.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+module.exports = class OsuRegister extends Command {
+  constructor (client) {
+    super(client, {
+      name: 'register',
+      aliases: ['reg', 'r'],
+      parentCommand: 'osu',
+      parameters: [{
+        type: 'string', full: true, missingError: 'commands:osu.subcommands.player.noPlayer'
+      }]
```
Comment (metehus) [inline src/commands/games/osu/register.js]: requeriment for database

**Comment by perronosaurio** [comment]
Idea: Add a top of all people with osu profiles registered in the bot that are on the server.
Example:
image

**Comment by dpaiv0** [comment]
what should we use for sorting out ppl? exp or pp?

**Comment by justandras** [comment]
Can't you use some kind of api to register users? This way they can register anyones account for themselves.

**Comment by pedrofracassi** [comment]
Yeah, unfortunately I believe there's no way for us to check if the account is theirs. Unless osu has some kind of profile bio that users can change and we can query thru the API

**Comment by dpaiv0** [comment]
there is [OAuth](https://github.com/int-and-his-friends/osu-api-v2/wiki/Authorization-code), we can use that

**Comment by pedrofracassi** [comment]
Cool. Let's use it.

**Comment by pedrofracassi** [comment]
@dpaiv0 how's osu oauth going?

**Comment by metehus** [comment]
connections!!

**Comment by dpaiv0** [comment]
osu! OAuth doesn't work.

**Comment by perronosaurio** [comment]
> osu! OAuth doesn't work.

why?

**Comment by dpaiv0** [comment]
when you try to get the correct code for everything to work, it just spills out wrong grant_type code

**Comment by pedrofracassi** [comment]
Maybe talk to @peppy and see if he can help?

**Comment by dpaiv0** [comment]
it seems to be a problem with Laravel, since OAuth is made using it

**Comment by peppy** [comment]
many projects are using it fine. please include code leading up to the error - you are likely doing something incorrectly.

**Comment by metehus** [comment]
you heard the guy you did shit

**Comment by dpaiv0** [comment]
@peppy sorry for wasting your time, it seems there was something wrong with the way I handled URLs, but it's now working, and retrieving every bit of data as it should be
image

File: src/apis/Osu.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  getAccessToken (authCode) {
+    return this.post('https://osu.ppy.sh/oauth/token', {
+      grant_type: 'authorization_code',
+      client_id: process.env.OSU_CLIENT_ID,
+      client_secret: process.env.OSU_CLIENT_SECRET,
+      redirect_uri: `${process.env.DASHBOARD_URL}/connections/osu/callback/`,
+      code: authCode
+    }).then(u => {
+      console.log(u)
```
Comment (perronosaurio) [inline src/apis/Osu.js]: why

File: src/commands/games/osu/player.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            t('commands:osu.subcommands.player.totalHits', { totalHits: MiscUtils.formatNumber((parseInt(userData.count300) + parseInt(userData.count100) + parseInt(userData.count50)), language), 300: MiscUtils.formatNumber(userData.count300, language), 100: MiscUtils.formatNumber(userData.count100, language), 50: MiscUtils.formatNumber(userData.count50, language), Constants })
+          ]
+        ]))
+
+      const topScores = await this.client.apis.osu.getUserTopScores(user, mode[0], 5)
+
+      if (topScores.length > 0) {
+        let description = []
+
+        for (var i in topScores) {
```
Comment (perronosaurio) [inline src/commands/games/osu/player.js]: hit or miss

File: src/locales/en-US/commands.json [hunk 5]
```diff
@@ -1059,6 +1059,43 @@
...
+        "commandDescription": "Searches for a beatmap or beatmap set on osu!",
+        "commandUsage": "<beatmap or beatmap set id>",
+        "noBeatmap": "You have to give me a valid beatmap or beatmap set ID.",
+        "starDifficulty": "Star Difficulty: **{{difficulty}}**",
+        "successRate": "Success Rate: **{{rate}}%** ({{successfulPlays}} / {{totalPlays}})",
+        "mappedBy": "Mapped by {{mapper}}",
+        "lastUpdate": "Last update on **{{date}}** ({{timeAgo}})",
+        "approved": "Approved on **{{date}}** ({{timeAgo}})",
+        "tags": "Tags: **{{tags}}**",
+        "download": "[Download]({{link}}) ([no video]({{linkNoVideo}}))",
```
Comment (perronosaurio) [inline src/locales/en-US/commands.json]: separe this with a -

File: src/commands/games/osu/player.js (hunk 4 above)
Comment (dpaiv0) [reply src/commands/games/osu/player.js]: ¿que?

**Comment by pedrofracassi** [comment]
@dpaiv0 	how's this going?

**Comment by dpaiv0** [comment]
@pedrofracassi I've yet tried to think what is still needed on this command. @perronosaurio any ideas?

**Comment by perronosaurio** [comment]
> I've yet tried to think what is still needed on this PR.

the tracking system

**Comment by pedrofracassi** [comment]
wdym tracking?

**Comment by metehus** [comment]
so.,......

**Comment by jaxxibae** [comment]
Closing this pull request, as it is from April 2019, and a lot of stuff changed. Will probably rewrite this today.

**Comment by jaxxibae** [comment]
(& also, the branch name, yuck)

**Pull request closed without merging**